### ISSUE TITLE
No way to configure AppMetrics Hosting Configuration from Startup

### ISSUE THREAD

**Issue by WillGunn** [AppMetrics/AppMetrics#255]
To the best of my knowledge, there is no way to set the url's for the metrics, text-metrics, and environment info endpoints in the startup class.  In version 2, they are only able to be set in the main entrypoint on WebHostBuilder.  This seems like a regression, as previously it could be set both from the startup class as well as from the appsettings.json file.

**Comment by alhardy** [comment]
It is still possible, apologies not in the docs. An approach could be as follows:

Startup.cs
```
public class Startup
    {
        public Startup(IConfiguration configuration) { Configuration = configuration; }

        public IConfiguration Configuration { get; }

        public void Configure(IApplicationBuilder app, IApplicationLifetime lifetime)
        {
            app.UseMetricsEndpoint();
            app.UseMetricsTextEndpoint();
            app.UseEnvInfoEndpoint();
            app.UseMetricsRequestTrackingMiddleware();
            app.UseMvc();
        }

        public void ConfigureServices(IServiceCollection services)
        {
            services.AddLogging();

            var metrics = AppMetrics.CreateDefaultBuilder().Build();
            services.AddMetrics(metrics);
            services.AddMetricsEndpoints();

            services.Configure<MetricsEndpointsHostingOptions>(Configuration.GetSection(nameof(MetricsEndpointsHostingOptions)));

            services.AddMvc().AddMetrics();
        }
```

appsettings.json

```
{
 ...
  "MetricsEndpointsHostingOptions": {
    "EnvironmentInfoEndpoint": "/custom-env",
    "MetricsEndpoint": "/custom-metrics",
    "MetricsTextEndpoint": "/custom-metrics-text" 
  } 
}
```

**Comment by WillGunn** [comment]
I’m pretty sure I tried that. I’ll check tomorrow though. The only other thing I could think that might’ve interfered was I had to create a custom reporter to hook into our logging system.

**Comment by WillGunn** [comment]
Yeah it wasn't the custom reporter, I commented that out and it still doesn't work even using
`services.Configure<MetricsEndpointsHostingOptions>(Configuration.GetSection(nameof(MetricsEndpointsHostingOptions)));
`

I'll post a slimmed down version of the setup I'm using (I had been using 1.0 but with all the changes in 2.0 re-went through the documentation setup guide to make sure I got everything).

Entrypoint: (if I uncomment out the lines below, the new urls works)
`			var host = new WebHostBuilder()
				.UseContentRoot(pathToContentRoot)
				.UseKestrelAndUrlsForEnvironment()
				//.UseMetrics()
				//.ConfigureAppMetricsHostingConfiguration(options =>
				//{
				//	options.MetricsEndpoint = "/priv/metrics";
				//	options.MetricsTextEndpoint = "/priv/metrics-text";
				//	options.EnvironmentInfoEndpoint = "/priv/environment-info";
				//})
				.UseStartup<Startup>()
				.Build();`
Startup:
`public class Startup
	{
		public IConfigurationRoot Configuration { get; }

		public Startup(IHostingEnvironment env)
		{
			var builder = new ConfigurationBuilder()
				.SetBasePath(env.ContentRootPath)
				.AddJsonFile("appsettings.json", optional: true, reloadOnChange: true)
				.AddJsonFile($"appsettings.{env.EnvironmentName}.json", optional: true)
				.AddEnvironmentVariables();
			Configuration = builder.Build();
		}`

Configure AppBuilder:
`		public void Configure(IApplicationBuilder app, ILoggerFactory loggerFactory, IServiceProvider serviceProvider)
		{
			SetupLoggerFactory(loggerFactory);

			// WARNING: The order of these statements directly affects the middleware processing order, from the top down.
			app.UseMetricsAllMiddleware();
			app.EnableCorsForApi(Configuration);
			AddBackupGlobalExceptionHandler(app);
			app.UseMvc();`

ConfigureServices:
`		public void ConfigureServices(IServiceCollection services)
		{
			services.Configure<MetricsEndpointsHostingOptions>(Configuration.GetSection(nameof(MetricsEndpointsHostingOptions)));
			//var metricsOptions = AppMetrics.CreateDefaultBuilder().Report.Using<HyperionMetricsReporter>(TimeSpan.FromSeconds(10)).Build();
			var metricsOptions = AppMetrics.CreateDefaultBuilder();
			//metricsOptions.Report.ToConsole(TimeSpan.FromSeconds(10));
			metricsOptions.Build();
			services.AddMetrics(metricsOptions);
			services.AddMetricsTrackingMiddleware();
			services.AddMetricsEndpoints();
			services.AddMetricsReportScheduler();
			services.AddMvc(AddMvcOptions).AddJsonOptions(options =>
			{
				options.SerializerSettings.Converters.Add(new ApiDateJsonConverter());
				options.SerializerSettings.Converters.Add(new StringEnumConverter());
			}).AddMetrics();`

appsettings.json:
`{
	"Logging": {
		"IncludeScopes": false,
		"LogLevel": {
			"Default": "Debug",
			"System": "Information",
			"Microsoft": "Information"
		}
	},
	"MetricsOptions": {
		"DefaultContextLabel": "redacted",
		"DefaultSamplingType": "ExponentiallyDecaying",
		"Enabled": true
	},
	"MetricsWebTrackingOptions": {
		"ApdexTrackingEnabled": true,
		"ApdexTSeconds": 0.2,
		"OAuth2TrackingEnabled": true
	},
	"MetricsEndpointsHostingOptions": {
		"EnvironmentInfoEndpoint": "/priv/environment-info",
		"MetricsEndpoint": "/priv/metrics",
		"MetricsTextEndpoint": "/priv/metrics-text"
	}
}
`

**Comment by alhardy** [comment]
@WillGunn is [this](https://github.com/AppMetrics/Samples.V2/commit/e4d388e8b65dd2cb92e557c34dfa1b55e2054a60) what you are after?

Having to configure the `MetricsEndpointHostingOptions` like [this](https://github.com/AppMetrics/Samples.V2/commit/e4d388e8b65dd2cb92e557c34dfa1b55e2054a60#diff-5e65153c8282c131ff54b4b60d727fd3R26) isn't ideal. Initially when `IWebHostBuilder` was provided by AspNet Core I was tempted to remove all the `IAppBuilder` extension methods assuming most would use `IWebHostBuilder` so missed this configuration option. I kept having the feeling that there are too many ways to bootstrap things, seems like this is the case :)

**Comment by WillGunn** [comment]
@alhardy Yeah.  In our solution we do all of our customization in the startup (IAppBuilder) area, instead of in the webhostbuilder.  

So looking at that gist, because we are bootstrapping it in the startup IAppBuilder area, do you not need the .UseMetrics() in IWebHostBuilder?

**Comment by alhardy** [comment]
Out of curiosity what makes you prefer to use `IApplicationBuilder` instead of `IWebHostBuilder`?

Correct, you only need to `UseMetrics()` on either interface. The `IWebHostBuilder` approach just simplifies the bootstrapping for you, it applies the [same extensions](https://github.com/AppMetrics/AspNetCore/blob/dev/src/App.Metrics.AspNetCore.Endpoints/DefaultMetricsEndpointsStartupFilter.cs#L21) on `IApplicationBuilder`, adds [required services](https://github.com/AppMetrics/AspNetCore/blob/dev/src/App.Metrics.AspNetCore/DefaultMetricsAspNetWebHostBuilderExtensions.cs#L25) etc.

The approach I've taken is similar to what `WebHost.CreateDefaultBuilder(args)` does, Serilog and other libraries do the same.

The main benefit I see is keeping infrastructure type configuration out of your `Startup.cs` which would be configuring more of your application specific "stuff".
